File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    # O(mn), O(mn)
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        maxarea = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    grid[i][j] = -1
                    queue = deque([(i, j)])
                    
                    area = 0
                    while queue:
                        for _ in range(len(queue)):
                            area += 1
                            
                            r, c = queue.popleft()
                            for x, y in directions:
                                nextr, nextc = r + x, c + y
                                if 0 <= nextr < rows and 0 <= nextc < cols and grid[nextr][nextc] == 1:
                                    grid[nextr][nextc] = -1
                                    queue.append((nextr, nextc))
                                    
                    maxarea = max(maxarea, area)
                    
        return maxarea

File: 0695-max-area-of-island/0695-max-area-of-island.cpp
/*
bfs on all ones
visited become 0
for each visit increment area
*/
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        rows = grid.size();
        cols = grid[0].size();
        
        int maxarea = 0;
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j]) {
                    grid[i][j] = 0;
                    maxarea = max(maxarea, bfs(grid, i, j));
                }
            }
        }
        
        return maxarea;
    }

private:
    int rows, cols;
    vector<pair<int, int>> directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    int bfs(vector<vector<int>>& grid, int starti, int startj) {
        queue<pair<int, int>> q;
        q.push({starti, startj});
        
        int area = 0;
        while (!q.empty()) {
            ++area;

            pair<int, int> land = q.front();
            q.pop();

            for (auto& direction : directions) {
                int newi = land.first + direction.first;
                int newj = land.second + direction.second;

                if (newi >= 0 && newi < rows && newj >= 0 && newj < cols && grid[newi][newj]) {
                    q.push({newi, newj});
                    grid[newi][newj] = 0;
                }
            }
        }
        
        return area;
    }
};